fl4health/strategies/fedpca.py

```diff
@@ -0,0 +1,269 @@
+                from clients that experienced an issue during training, such as timeouts or exceptions.
+
+        Returns:
+            Tuple[Optional[Parameters], Dict[str, Scalar]]: The aggregated parameters and the metrics dictionary.
+                In this case, the parameters are the new singular vectors and their corresponding singular values.
+        """
+        if not results:
+            return None, {}
+        # Do not aggregate if there are failures and failures are not accepted
+        if not self.accept_failures and failures:
+            return None, {}
+
+        client_singular_values = []
+        client_singular_vectors = []
+        for _, fit_res in results:
+            A = parameters_to_ndarrays(fit_res.parameters)
+            singular_vectors, singular_values = A[0], A[1]
+            client_singular_vectors.append(singular_vectors)
+            client_singular_values.append(singular_values)
+
+        if self.svd_merging:
+            log(INFO, "Performing SVD-based merging.")
+            merged_singular_vectors, merged_singular_values = self.merge_subspaces_svd(
+                client_singular_vectors, client_singular_values
+            )
+        else:
+            # use qr merging instead
+            log(INFO, "Performing QR-based merging.")
+            merged_singular_vectors, merged_singular_values = self.merge_subspaces_qr(
+                client_singular_vectors, client_singular_values
+            )
+        parameters = ndarrays_to_parameters([merged_singular_vectors, merged_singular_values])
+
+        # Aggregate custom metrics if aggregation fn was provided
+        metrics_aggregated = {}
+        if self.fit_metrics_aggregation_fn:
+            fit_metrics = [(res.num_examples, res.metrics) for _, res in results]
+            metrics_aggregated = self.fit_metrics_aggregation_fn(fit_metrics)
+        elif server_round == 1:  # Only log this warning once
+            log(WARNING, "No fit_metrics_aggregation_fn provided")
+
+        return parameters, metrics_aggregated
+
+    def merge_subspaces_svd(
+        self, client_singular_vectors: NDArrays, client_singular_values: NDArrays
+    ) -> Tuple[NDArray, NDArray]:
+        """
+        Produce the principal components (PCs) for all the data distributed across clients by merging the PCs
+        belonging to each local dataset.
+
+        Each clients sends a matrix whose columns are the local principal components to the server. The corresponding
+        singular values are also shared.
+
+        The server then arranges the local PCs into a block matrix, then perform SVD.
+
+        For example, if U_i denotes the matrix of PCs and S_i denotes the diagonal matrix of
+        singular values for client i, and there are n clients, then merging is done by
+        performing SVD on the matrix
+
+        B = [U_1 @ S_1 | U_2 @ S_2 | ... | U_n @ S_n],
```

Comment:
I think the more standard notation here would be `diag([U_1 @ S_1, U_2 @ S_2, ..., U_n @ S_n])`, but perhaps I'm wrong.

Reply:
I meant for this to mean that `B` is a block matrix consisting of blocks `U1S1, U2S2, ...,  UnSn` from arranged from left to right. I wasn't sure what a good notation for this would be in the docstring

Reply:
Gotcha. So if this is supposed to be a block diagonal matrix, I think a standard notation for this would be similar to the diag notation I put above. That is, if n = 2, you want to indicate a matrix that looks like this
```
U_1 @ S_1           0
      0            U_2 @ S_2
```
where the zeros are blocks of 0s with the appropriate dimensions.

Reply:
Actually B wouldn't be a block diagonal matrix. It would just be a matrix with blocks `U_1 @ S_1` and `U_2 @ S_2` placed from left to right (i.e., there won't be any zero blocks).

So for example, if `U_1 @ S_1` has size d by n1 and `U_2 @ S_2` has size d by n2, then `B` would have size d by (n1 + n2).

Reply:
Ah okay I see. Your comments are clear on this. I'm not sure why I assume we needed it to be block diagonal. Apologies.

Reply:
No worries at all! The notation was a bit confusing